### ISSUE TITLE
Stopping via `max_tokens` should include captures if in a matched state

### ISSUE THREAD

**Issue by hudson-ai** [hudson-ai/guidance#21]
**TLDR**: stopping generation when we hit `max_tokens` should include matches if "properly" stopping at this point (e.g. via an EOS) would include matches. Related to issue #15.

Below is a bit verbose and meanders its way to the central issue. This is only included for historical reasons ;)

**The bug**
```python
import re
from guidance import models, gen

model = models.Transformers("microsoft/Phi-3-mini-4k-instruct")

prompt = "2 + 2 = "
pattern = r"\d+\.\d+"

m = model + prompt + gen(name="test", regex=pattern, max_tokens=40)
assert re.fullmatch(pattern, m["test"])

>> KeyError: "Model does not contain the variable 'test'"
```

What's going on here...?
```python
str(m)
>> '2 + 2 = 4.000000000000002᠐᠓᠑᠐᠓᠑᠐'
```
We generated some weird stuff. *Technically*, this stuff matches our espression, so maybe this isn't necessarily related to the problem. The only way out of the gen expression is an EOS, which may somehow be less likely than these weird characters (yes, changing the stop condition helps... adding decimals to the prompt also helps).

But why aren't we capturing the output? We actually DO have a match...
```python
re.fullmatch(pattern, '4.000000000000002᠐᠓᠑᠐᠓᠑᠐') is not None
>> True
```

Clearly, we hit `max_tokens` and didn't emit a capture despite being in a "matched" state.

**Comment by hudson-ai** [comment]
@mmoskal I take this back -- I don't think my original title nor message captures the underlying issue. If that were the case, we wouldn't capture anything with 
```python
m = model + gen(max_tokens=3, name='gen')
```
(which is just fine)

**Comment by hudson-ai** [comment]
Furthermore, changing the prompt to `"2.0 + 2.0 = "` yields 
```python
str(m)
>> '2.0 + 2.0 = 4.00000000000000000000000000000000000000'

m["test"]
>> '4.00000000000000000000000000000000000000'
```
which is also behaving as expected. The long tail of zeroes *feels* odd, but it goes away if you give the gen a stop of `\n`, so it's just an artifact of waiting for an EOS and not getting it...

Maybe the weird unicode digits ARE to blame?

**Comment by mmoskal** [comment]
So what is the repro?

```python
import os
import guidance
from guidance import gen


def main():
    os.environ["LLGUIDANCE_LOG_LEVEL"] = "10"
    lm = guidance.models.LlamaCpp(
        "/Users/mimoskal/.cache/huggingface/hub/models--TheBloke--Llama-2-7B-GGUF/snapshots/b4e04e128f421c93a5f1e34ac4d7ca9b0af47b80/llama-2-7b.Q5_K_M.gguf"
    )
    lm += "2 + 2 = " + gen(name="test", regex=r"\d+\.\d+", max_tokens=8)
    print(lm["test"])


main()
```

This seems to work for various values of 8

**Comment by hudson-ai** [comment]
@mmoskal I think the phi3 tokenizer is to blame. I know there is some WEIRD stuff lurking there.

Try the first code block I sent, where LlamaCpp = Transformers, TheBloke--Llama-2-7B-GGUF = microsoft/Phi-3-mini-4k-instruct, and 8 = 40

**Comment by hudson-ai** [comment]
Ok, tracked down the issue. For the phi tokenizer above, the non-ASCII digits are multi-token digits (corresponding to the individual bytes of some three-byte unicode codes). The token limit just so happens to stop in the middle of a multi-token digit here, causing the parser to be in a non-accepting state and preventing a capture.

In all, I would say that this isn't really a bug, just somewhat confusing behavior. Tests weren't hitting this on main due to the old regex implementation not supporting general unicode. I'll call this a test misspecification and close the issue.

**Comment by hudson-ai** [comment]
Note that tracked issue #23 would at least make this behavior less confusing by ensuring that it isn't happening silently.
